config_v2.md

```diff
@@ -0,0 +1,370 @@
+# Sacred config v2
+
+## Goals of this document:
+
+* Present the API for the new way to work with sacred configs
+
+* Make sure this API can handle all the use case that used to work before.
+
+* Make sure that this API is intuitive and Pythonic.
+
+* Make sure that someone who has never read the sacred docs can understand some codebase that uses sacred.
+
+* Remove all black magic of configs. It is linked to the previous point.
+
+* Give enough flexibility to allow implementation of many of the features requests about configs in github issues.
+
+## Non-goals of this document:
+
+* Describing or even caring about the migration process. It can be done at a later stage.
+
+* Caring about looking like our old API about configs. Legacy API design should not prevent us from pushing a better API.
+
+* Handling seeding. The seeding has been separated from the configuration process with the PR [Making the seeding more intuitive](https://github.com/IDSIA/sacred/pull/615)
+
+
+## API description:
+
+The goal is to make the API easier to understand.
+
+* There is no more `Ingredient`. 
+
+* There is no config scopes.
+
+* Ingredients are replaced by functions which can be activated either from the command line or when creating the Experiment.
+
+* No local variable gathering, injections or override. In short, no black magic.
+
+* Classes and functions can be put in the config, but only the full name will be saved with the Observers.
+
+* We have a Config object now, and all the command lines updates are done at the end of the constructor. The Experiment does not modify the Config object.
+
+### Basic example:
+
+
+```python
+import sacred
+from sacred import Config
+
+
+configuration = Config(dict(batch_size=32, dataset_size=10_000, nb_epochs=50))
+
+ex = sacred.Experiment('my_pretty_experiment', config=configuration)
+                       
+def my_main_function(batch_size, dataset_size, nb_epochs):
+    # main experiment here
+    ...
+
+with ex.start():
```

Comment:
Currently, I can start an experiment explicitly via `ex.run` and also indirectly via `ex.run_commandline()`. I find both these options convenient -- e.g. running multiple experiments with different parameters via `ex.run`, and stand-alone CLI scripts using `ex.run_commandline` -- and often use both for the same experiment. Can this use-case be supported here?

Reply:
For sure. We want to support use cases that worked before as much as possible. This is very important. Could you help us by providing a dummy Gist with what you describe? It doesn't have to work, it's for us to understand your use case better and make sure we can support it. I understand the big picture of what you're describing but code is much better to make sure we don't misunderstand your use case.

Reply:
Here's a gist illustrating this use case in the context of a hyperparameter sweep: https://gist.github.com/AdamGleave/ce76233da6f4449e53c788f3da02ed7d

Suppose I wrote complex training script `standalone.py`. Initially I just run it by itself as a standalone experiment. At some point I want to do a hyperparameter sweep, so I write a second script `sweep.py` that runs the experiment in `standalone.py` using `standalone_ex.run`. I still want to be able to use `standalone.py` directly, via `standalone_ex.run_commandline()`. In practice `sweep.py` would often be spawning multiple processes, possibly on different machines, each executing `standalone_ex.run`.

Reply:
Thanks a lot for the gist! It's really helpful, I understand much better now. Give me some time to adapt my proposal to your use case.

Reply:
Thanks for the update, I think this supports my use case now :)

Reply:
Cool! Thanks for the feedback!